data_structures/linked_list/floyds_cycle_detection.py

```diff
@@ -0,0 +1,150 @@
+"""
+Floyd's cycle detection algorithm is a popular algorithm used to detect cycles
+in a linked list. It uses two pointers, a slow pointer and a fast pointer,
+to traverse the linked list. The slow pointer moves one node at a time while the fast
+pointer moves two nodes at a time. If there is a cycle in the linked list,
+the fast pointer will eventually catch up to the slow pointer and they will
+meet at the same node. If there is no cycle, the fast pointer will reach the end of
+the linked list and the algorithm will terminate.
+
+For more information: https://en.wikipedia.org/wiki/Cycle_detection#Floyd's_tortoise_and_hare
+"""
+
+from collections.abc import Iterator
+from dataclasses import dataclass
+from typing import Any, Self
+
+
+@dataclass
+class Node:
+    """
+    A class representing a node in a singly linked list.
+    """
+
+    data: Any
+    next_node: Self | None = None
+
+
+@dataclass
+class LinkedList:
+    """
+    A class representing a singly linked list.
+    """
+
+    head: Node | None = None
+
+    def __iter__(self) -> Iterator:
+        """
+        Iterates through the linked list.
+
+        Returns:
+            Iterator: An iterator over the linked list.
+
+        Examples:
+        >>> linked_list = LinkedList()
+        >>> list(linked_list)
+        []
+        >>> linked_list.add_node(1)
+        >>> tuple(linked_list)
+        (1,)
+        """
+        visited = []
+        node = self.head
+        while node:
+            # Avoid infinite loop in there's a cycle
+            if node in visited:
+                return
+            visited.append(node)
```

Comment:
```suggestion
        visited = set()
        node = self.head
        while node:
            # Avoid infinite loop in there's a cycle
            if node in visited:
                return
            visited.add(node)
```
A set makes more sense for this purpose, IMO

Reply:
Nevermind, forgot that the Nodes aren't hashable